crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,428 @@
+/// work loop that gathers checkpoint-wise row information, and periodically writes them out to the
+/// database.
+///
+/// The period between writes is controlled by the following factors:
+///
+/// - Time since the last write (controlled by `config.commit_interval`). If there are rows pending
+///   and this interval has elapsed since the last attempted commit, the committer will attempt
+///   another write.
+///
+/// - Time since last attempted write (controlled by `COOLDOWN_INTERVAL` and `RETRY_INTERVAL`). If
+///   there was a recent successful write, the next write will wait at least `COOLDOWN_INTERVAL`,
+///   and if there was a recent unsuccessful write, the next write will wait at least
+///   `RETRY_INTERVAL`. This is to prevent one committer from hogging the database.
+///
+/// - Number of pending rows. If this exceeds `H::BATCH_SIZE`, the committer will attempt to write
+///   out at most `H::CHUNK_SIZE` worth of rows to the DB.
+///
+/// If a write fails, the commiter will save the batch it meant to write and try to write them
+/// again at the next opportunity, potentially adding more rows to the batch if more have arrived
+/// in the interim.
+///
+/// This task will shutdown if canceled via the `cancel` token, or if the channel it receives data
+/// on has been closed by the handler for some reason.
+fn committer<H: Handler + 'static>(
+    db: Db,
+    mut rx: mpsc::Receiver<Indexed<H>>,
+    config: CommitterConfig,
+    metrics: Arc<IndexerMetrics>,
+    cancel: CancellationToken,
+) -> JoinHandle<()> {
+    spawn_monitored_task!(async move {
+        // The `poll` interval controls the maximum time to wait between commits, regardless of the
+        // amount of data available.
+        let mut poll = interval(config.commit_interval);
+        let mut cool = interval(COOLDOWN_INTERVAL);
+
+        // We don't care about keeping a regular cadence -- these intervals are used to guarantee
+        // things are spaced at out relative to each other.
+        poll.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        cool.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        // Buffer to gather the next batch to write. This may be non-empty at the top of a tick of
+        // the commiter's loop if the previous attempt at a write failed. Attempt is incremented
+        // every time a batch write fails, and is reset when it succeeds.
+        let mut attempt = 0;
+        let mut batch = vec![];
+
+        // Data for checkpoints that haven't been written yet. Note that `pending_rows` includes
+        // rows in `batch`.
+        let mut pending: BTreeMap<u64, Vec<H::Value>> = BTreeMap::new();
+        let mut pending_rows = 0;
+
+        // TODO: Track the watermark (keep track of out-of-order commits, and update the watermark
+        // table when we accumulate a run of contiguous commits starting from the current watermark).
+
+        loop {
```

Comment:
This loop is fairly complex and subtle — it took me a few reads to fully understand it. If we ever need to add more branches to the select statement in the future (even though I don’t foresee it now), there’s a risk of introducing deadlocks or similar issues, especially if the intervals aren’t configured properly, causing the loop to always enter a particular branch of the select (due to `biased`) even when other conditions are met and should be executed.

I don't have any concrete suggestions right now on how to make this less error prone but wondering if you have any thoughts.

Reply:
It is -- I don't have a great suggestion for how to make this particular part less complex, but I'd say this is about the limit of complexity we want in a single task, and if we need to do more, we should consider factoring things out into separate tasks (I don't think there's an easy way to do that for this though, there is some inherent complexity here).

Some of the complexity comes from wanting to make sure we can test out some hypotheses (like whether we need to have a cooldown in the first place).

Reply:
> Likelihood of deadlock is fairly unlikely though, for what it's worth, because most of the cases in the select are timers

We can get into this situation if one interval is shorter than the other one and is written first in the branch, then we may always hit this branch even though the other way is also ready because of `biased`. Like this code:
```rust
loop {
    select! {
        biased;

        one_second_interval.tick() => {
            do_something_taking_more_than_one_second(); 
        }
        two_second_interval.tick() => {
            // This will never be reached.
        }     
    }
}
```
That being said, I don't think the current code has this issue. Just pointing out that, even with mostly timer cases, there's still risk and we should be careful with choosing the interval time as well as the order of branches in `select`, as we add more cases.

Reply:
Yeah, point taken -- we can also consider removing the `biased;` if we find that it isn't adding much value. That's another worthwhile experiment! I'll add a TODO for that.

Your example is also what I meant by this:

> but the thing we might need to worry about is setting the appropriate guards on the selects so that we only query the timer when the work it's guarding can actually be done.

That issue only shows up if the futures passed to `select!` are unguarded (like in your example), but if they are appropriately guarded, it won't happen because the guards are used to filter which futures are polled on a given iteration, so in this example:

```rust
let mut let_me_through = false;
loop {
  select! {
    biased;

    one_second_interval.tick(), if let_me_through => {
      /* never reached */
    }

    two_second_interval.tick() => {
      /* do something */
    }
  }
}
```

Even though the one second interval fires first, we never evaluate its arm because of the guard.

Reply:
Added an experimentation comment so we don't forget to test out whether we can remove `biased` which may significantly reduce the subtlety here (I reckon we probably can remove it).